### DISCUSSION TITLE
[✨ feature request] Is "change of emitter during rendering"available?

### DISCUSSION THREAD

**Discussion by zhaoguangyuan123** [mitsuba-renderer/mitsuba3#197] [Feature requests]
## Summary

Just want this simple feature. 
```python
init_images = [mi.render(scene, emitter=emitters[i], spp=128) for i in range(emitter_count)]
```
From the document and tutorials I can only see that the "sensor" but not the "emitter" is able to change. 
Is such feature already available or any other workarounds??

Looking forward to hearing any suggestions from you.

Best
Guangyuan

**Comment by merlinND** [comment]
Hello @zhaoguangyuan123,

Such an argument in the `mi.render()` interface wouldn't make a lot of sense, because a scene can have an arbitrary amount of emitters.
Instead, you could e.g. create a scene with all emitters in it and use the `mi.traverse()` mechanism to update the value of the emitters before rendering. Something along the lines of:

```python
params = mi.traverse(scene)

def enable_only_emitter(name):
    for k, _ in params.items():
        if k == name + '.radiance.value':  # Check exactly how the relevant property is named in your scene
            params[k] = 1.5  # Will also depend on your scene
        else:
            params[k] = 0.
    params.update()

for i in range(emiter_count):
    enable_only_emitter('emitter_' + i)
    img = mi.render(scene, spp=128)
    # ...
```
Of course you'll have to adapt the naming convention and the parameter data types according to what's in your scene.

Another option is to create the scene within each loop iteration using `load_dict` and giving only the emitter you want in that iteration.

**Reply by zhaoguangyuan123** [reply]
Hi Merlin,

Appreciate it a lot for your kind reply during Sunday (, which I did not expect).

I am afraid the solution above does not answer my question. 

I want the emitter to be replaced, but not updated in every rendering. 
The emitters here is from a dataset; I do not want to update the parameter of emitter but want a single emitter to present in the scene in every rendering. 
In other words, the emitters in the dataset are exclusive to each other.  In such way, the use of emitter is similar to the 'sensor' in your example of "MI-NERF".

Pls kindly let me know if my understanding is incorrect.


Best
Guangyuan

**Reply by zhaoguangyuan123** [reply]
> Another option is to create the scene within each loop iteration using `load_dict` and giving only the emitter you want in that iteration.

Would this option solve my issue？
My concern is that in every 'load_dict', the parameters, such as the 'lens' in the dataset is not kept but re-initialized. 

Best,

**Reply by merlinND** [reply]
Hi @zhaoguangyuan123,

It's true that reloading the scene would be inefficient at runtime (re-created all objects every time), but it would be the easiest to implement.

What type of emitter are you using in the scene? (`point`, `area`, something else?)
`params` is very flexible, so even if you just create one emitter at the start when you load the scene, you could also change the position (`to_world`) and intensity of the emitter to where it needs to be just before making each rendering.

**Reply by zhaoguangyuan123** [reply]
> 

So I am basically trying to introduce computer vision tasks into Mitsuba. The easiest way to introduce dataset (such as ImageNet) I can think is to make the source images as emitters in the 'scene'.
In other words, the **area** emitter in every rendering is an instance from ImageNet.

Is that feasible?

**Reply by merlinND** [reply]
I am not sure I understand the purpose of having ImageNet images as emitters in a 3D scene.
What are the other objects in the scene, should they be illuminated by that emitter?
If there is no other object, then is rendering even needed? (could work with the image itself directly in 2D instead)

Anyway, yes it would be achievable by using a textured [AreaLight](https://mitsuba.readthedocs.io/en/latest/src/generated/plugins_emitters.html#area-light-area) emitter.
At each iteration, you could use `params` to set the values of the emissive texture to the new image from ImageNet.

**Reply by zhaoguangyuan123** [reply]
> I am not sure I understand the purpose of having ImageNet images as emitters in a 3D scene. What are the other objects in the scene, should they be illuminated by that emitter? If there is no other object, then is rendering even needed? (could work with the image itself directly in 2D instead)
> 
> Anyway, yes it would be achievable by using a textured [AreaLight](https://mitsuba.readthedocs.io/en/latest/src/generated/plugins_emitters.html#area-light-area) emitter. At each iteration, you could use `params` to set the values of the emissive texture to the new image from ImageNet.

Sure. Thank you.

You are right that ImageNet is only 2D. Ideally, I want  to load 3D dataset but my thought was to start with ImageNet, as the 2D object is relatively easier. 

I just come up with another possible way. For the same task, can I just initialize a object in between the **light source** and **sensor** the  while replace the object in the scene in every rendering and keep the emitter to be fixed?

**Reply by zhaoguangyuan123** [reply]
> What are the other objects in the scene, should they be illuminated by that emitter?

Yes in my previous plan the other objects in the scene are the object to optimize for and they are also illuminated by the emitter. 
Let me know if this does not resolve your confusion.

**Reply by merlinND** [reply]
> replace the object in the scene in every rendering

Unfortunately, that's actually not supported at the moment, we cannot insert or remove shapes once a scene has been loaded.

> the other objects in the scene are the object to optimize for and they are also illuminated by the emitter.

Okay, in that case I think that the textured area emitter would work!

**Reply by zhaoguangyuan123** [reply]
> 

Sure. I will go ahead and try the "Arealight" plan you suggest.
Thank you, Merlin!
